core_backend/tests/rails/test_llm_response_in_context.py

```diff
@@ -3,30 +3,45 @@
 LLM response content validation functions are rails that check if the responses
 are based on the given context or not.
 """
-from typing import Dict, List, Literal
+from pathlib import Path
+from typing import List, Literal, Tuple
 
 import pytest
 import yaml
 
+from core_backend.app.configs.app_config import ALIGN_SCORE_API, ALIGN_SCORE_THRESHOLD
+from core_backend.app.llm_call.check_output import _get_alignScore_score
+
 pytestmark = pytest.mark.rails
 
-CONTEXT_RESPONSE_FILE = "tests/rails/data/llm_response_in_context.yaml"
+CONTEXT_RESPONSE_FILE = "data/llm_response_in_context.yaml"
 TestDataKeys = Literal["context", "statement", "expected", "reason"]
 
 
-def read_test_data(file: str) -> List[Dict[TestDataKeys, str]]:
+def read_test_data(file: str) -> List[Tuple]:
     """Reads test data from file and returns a list of strings"""
 
-    with open(file, "r") as f:
+    file_path = Path(__file__).parent / file
+
+    with open(file_path, "r") as f:
         content = yaml.safe_load(f)
-        return [dict(context=c["context"], **t) for c in content for t in c["tests"]]
+        return [(c["context"], *t.values()) for c in content for t in c["tests"]]
 
 
-def test_alignScore(llm_sentence: str, context: str, expected_answer: bool) -> None:
+@pytest.mark.asyncio(scope="module")
+@pytest.mark.parametrize(
+    "context, statement, expected, reason", read_test_data(CONTEXT_RESPONSE_FILE)
+)
+async def test_alignScore(
```

Comment:
When I run the test command given in the PR description (after having activated a fresh install environment and setting OpenAI key) I get the following error:

```
=========================================================== warnings summary ============================================================
../../../../../opt/homebrew/lib/python3.11/site-packages/_pytest/config/__init__.py:1373
  /opt/homebrew/lib/python3.11/site-packages/_pytest/config/__init__.py:1373: PytestConfigWarning: Unknown config option: asyncio_mode
  
    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

tests/api/test_question_answer.py:157
  /Users/amiremami/IDinsight Repos/aaq-core/core_backend/tests/api/test_question_answer.py:157: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio

tests/api/test_question_answer.py:176
  /Users/amiremami/IDinsight Repos/aaq-core/core_backend/tests/api/test_question_answer.py:176: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio

tests/rails/test_llm_response_in_context.py:31
  /Users/amiremami/IDinsight Repos/aaq-core/core_backend/tests/rails/test_llm_response_in_context.py:31: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio(scope="module")

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================ 48 skipped, 122 deselected, 4 warnings in 2.12s ============================================
```

Reply:
Sorry. Forgot to add this to `requirements-dev.txt`
```
pytest-asyncio==0.23.2
```

Adding now but you can install it in the meantime.

Reply:
Thanks. Now tests are not skipped but failing with the error following error. Can you try the PR steps using a fresh environment (fresh env, set openai key, run that pytest command) to check if you get the same issue? I'm making my new env using `make fresh-env`.

```
self = <aiohttp.connector.TCPConnector object at 0x151a39bd0>, req = <aiohttp.client_reqrep.ClientRequest object at 0x155955390>
timeout = ClientTimeout(total=300, connect=None, sock_read=None, sock_connect=None, ceil_threshold=5)
client_error = <class 'aiohttp.client_exceptions.ClientConnectorError'>
args = (functools.partial(<class 'aiohttp.client_proto.ResponseHandler'>, loop=<_UnixSelectorEventLoop running=False closed=False debug=False>), '::1', 5001)
kwargs = {'family': <AddressFamily.AF_INET6: 30>, 'flags': <AddressInfo.AI_NUMERICHOST|AI_NUMERICSERV: 4100>, 'local_addr': None, 'proto': 6, ...}

    async def _wrap_create_connection(
        self,
        *args: Any,
        req: ClientRequest,
        timeout: "ClientTimeout",
        client_error: Type[Exception] = ClientConnectorError,
        **kwargs: Any,
    ) -> Tuple[asyncio.Transport, ResponseHandler]:
        try:
            async with ceil_timeout(
                timeout.sock_connect, ceil_threshold=timeout.ceil_threshold
            ):
                return await self._loop.create_connection(*args, **kwargs)
        except cert_errors as exc:
            raise ClientConnectorCertificateError(req.connection_key, exc) from exc
        except ssl_errors as exc:
            raise ClientConnectorSSLError(req.connection_key, exc) from exc
        except OSError as exc:
            if exc.errno is None and isinstance(exc, asyncio.TimeoutError):
                raise
>           raise client_error(req.connection_key, exc) from exc
E           aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host localhost:5001 ssl:default [Connect call failed ('::1', 5001, 0, 0)]

/opt/homebrew/lib/python3.11/site-packages/aiohttp/connector.py:1000: ClientConnectorError
```

Reply:
48 failed, 122 deselected in 6.09s

Reply:
Ran the pytests with the new instructions from the PR and now getting:

```
/opt/homebrew/lib/python3.11/site-packages/aiohttp/streams.py:622: ClientOSError
_ test_alignScore[Full blood count testing is the recommended method for diagnosing\nanaemia during pregnancy. In settings where full blood count testing is not\navailable, onsite haemoglobin testing with a haemoglobinometer is recommended over\nthe use of the haemoglobin colour scale as the method for diagnosing anaemia in\npregnancy. Midstream urine culture is the recommended method for diagnosing\nasymptomatic bacteriuria (ASB) in pregnancy. In settings where urine culture is not\navailable, on-site midstream urine Gram-staining is recommended over the use of\ndipstick tests as the method for diagnosing ASB in pregnancy. Clinical enquiry about\nthe possibility of intimate partner violence (IPV) should be strongly considered at\nantenatal care visits when assessing conditions that may be caused or complicated by\nIPV in order to improve clinical diagnosis and subsequent care, where there is the\ncapacity to provide a supportive response (including referral where appropriate) and\nwhere the WHO minimum requirements are met. Hyperglycaemia first detected at any time\nduring pregnancy should be classified as either gestational diabetes mellitus (GDM)\nor diabetes mellitus in pregnancy, according to WHO criteria. Health-care providers\nshould ask all pregnant women about their tobacco use (past and present) and exposure\nto second-hand smoke as early as possible in pregnancy and at every antenatal care\nvisit. Health-care providers should ask all pregnant women about their\nuse of alcohol and other substances (past and present) as early as possible in the\npregnancy and at every antenatal care visit.\n-Doctors need to ask about smoking, alcholol, and substance during every visit\n-True-They should be asking these at every visit] _

context = 'Full blood count testing is the recommended method for diagnosing\nanaemia during pregnancy. In settings where full b...ol and other substances (past and present) as early as possible in the\npregnancy and at every antenatal care visit.\n'
statement = 'Doctors need to ask about smoking, alcholol, and substance during every visit\n', expected = True
reason = 'They should be asking these at every visit'

    @pytest.mark.asyncio(scope="module")
    @pytest.mark.parametrize(
        "context, statement, expected, reason", read_test_data(CONTEXT_RESPONSE_FILE)
    )
    async def test_alignScore(
        context: str, statement: str, expected: bool, reason: str
    ) -> None:
        """
        This checks if alignScore returns the correct answer
        """
>       score = await _get_alignScore_score(
            ALIGN_SCORE_API, {"evidence": context, "claim": statement}
        )

tests/rails/test_llm_response_in_context.py:41: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
app/llm_call/check_output.py:101: in _get_alignScore_score
    async with get_http_client().post(api_url, json=payload) as resp:
/opt/homebrew/lib/python3.11/site-packages/aiohttp/client.py:1187: in __aenter__
    self._resp = await self._coro
/opt/homebrew/lib/python3.11/site-packages/aiohttp/client.py:601: in _request
    await resp.start(conn)
/opt/homebrew/lib/python3.11/site-packages/aiohttp/client_reqrep.py:965: in start
    message, payload = await protocol.read()  # type: ignore[union-attr]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <aiohttp.client_proto.ResponseHandler object at 0x149152900>

    async def read(self) -> _T:
        if not self._buffer and not self._eof:
            assert not self._waiter
            self._waiter = self._loop.create_future()
            try:
>               await self._waiter
E               aiohttp.client_exceptions.ClientOSError: [Errno 54] Connection reset by peer

/opt/homebrew/lib/python3.11/site-packages/aiohttp/streams.py:622: ClientOSError
```

Can you delete all docker images + remove temp files + fresh env and see if it runs for you if you follow the instructions from scratch?

Reply:
It may have been the healthcheck on the AlignScore container. I pushed that change just in case it was.

Then ran the following:

```
docker ps -aq | xargs docker stop | xargs docker rm 
docker container prune
docker image prune -a
docker volume prune

cd core_backend
make fresh-env
 
conda activate aaq-core

cd optional_components/alignScore
docker build -t alignscore-base .
docker run -p 5001:5001 --name align-score-local --detach 'alignscore-base'
cd ../../core_backend
pytest -m rails -k "test_alignScore" -vv
```

And it runs!